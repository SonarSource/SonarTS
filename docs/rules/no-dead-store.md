# no-dead-store

A dead store happens when a local variable is assigned a value that is not read by any subsequent instruction or when an object property is assigned a value that is not subsequently used. Calculating or retrieving a value only to then overwrite it or throw it away, could indicate a serious error in the code. Even if it's not an error, it is at best a waste of resources. Therefore all calculated values should be used.

## Noncompliant Code Example

```typescript
function pow(a, b) {
  if(b == 0) {
    return 0;
  }
  var x = a;
  for(var i = 1; i < b; i++) {
    x = x * a;  //Dead store because the last return statement should return x instead of returning a
  }
  return a;
}
```

## Compliant Solution
```typescript
function pow(a, b) {
  if(b == 0) {
    return 0;
  }
  var x = a;
  for(var i = 1; i < b; i++) {
    x = x * a;
  }
  return x;
}
```

## Exceptions
This rule ignores initializations to `-1`, `0`, `1`, `null`, `true`, `false`, `""`, `[]` and `{}`.
