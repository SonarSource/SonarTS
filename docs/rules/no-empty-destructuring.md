# no-empty-destructuring

Destructuring is a convenient way of extracting multiple values from data stored in 
(possibly nested) objects and arrays. However, it is possible to create an empty pattern that has 
no effect. When empty curly braces or brackets are used to the right of a property name most of the 
time the intent was to use a default value instead.

This rule raises an issue when empty destructuring pattern is used.

## Noncompliant Code Example

```javascript
var { a: {}, b } = myObj; // Noncompliant
function foo({ first: [], second }) { // Noncompliant
  // ...
}
```

## Compliant Solution

```javascript
var { a = {}, b } = myObj;
function foo({ first = [], second }) {
  // ...
}
```
