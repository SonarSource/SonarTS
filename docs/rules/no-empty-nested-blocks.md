# no-empty-nested-blocks

Most of the time a block of code is empty when a piece of code is really missing. So such empty block must be either 
filled or removed.

##  Noncompliant Code Example

```typescript
for (let i = 0; i < 42; i++) {}  // Empty on purpose or missing piece of code ?
```

## Exceptions

When a block contains a comment, this block is not considered to be empty. Moreover `catch` blocks are ignored.
