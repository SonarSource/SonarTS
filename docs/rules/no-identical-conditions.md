# no-identical-conditions

A `switch` and a chain of `if`/`else if` statements is evaluated from top to bottom. At most, only 
one branch will be executed: the first one with a condition that evaluates to `true`.

Therefore, duplicating a condition automatically leads to dead code. Usually, this is due to a 
copy/paste error. At best, it's simply dead code and at worst, it's a bug that is likely to induce 
further bugs as the code is maintained, and obviously it could lead to unexpected behavior.

For a `switch`, if the first case ends with a `break`, the second case will never be executed, 
rendering it dead code. Worse there is the risk in this situation that future maintenance will be 
done on the dead case, rather than on the one that's actually used.

On the other hand, if the first case does not end with a `break`, both cases will be executed, 
but future maintainers may not notice that.

## Noncompliant Code Example

```typescript
if (param == 1)
  openWindow();
else if (param == 2)
  closeWindow();
else if (param == 1)  // Noncompliant
  moveWindowToTheBackground();

switch (i) {
  case 1:
    //...
    break;
  case 3:
    //...
    break;
  case 1:  // Noncompliant
    //...
    break;
  default:
    // ...
    break;
}
```