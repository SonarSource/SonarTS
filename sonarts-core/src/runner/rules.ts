/*
 * SonarTS
 * Copyright (C) 2017-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as path from "path";
import * as ts from "typescript";
import * as tslint from "tslint";

const SONARTS_RULES_FOLDER = path.join(__dirname, "../../lib/rules");

export function getIssues(
  ruleConfigs: tslint.IOptions[],
  program: ts.Program,
  sourceFile: ts.SourceFile,
): { issues: any[] } {
  const rules = tslint.loadRules(ruleConfigs, SONARTS_RULES_FOLDER);
  let issues: tslint.RuleFailure[] = [];
  rules.forEach(rule => (issues = issues.concat(executeRule(rule, sourceFile, program))));
  return { issues: issues.map(issue => issue.toJson()) };
}

export function executeRule(rule: tslint.IRule, sourceFile: ts.SourceFile, program: ts.Program): tslint.RuleFailure[] {
  try {
    return isTypedRule(rule) ? rule.applyWithProgram(sourceFile, program) : rule.apply(sourceFile);
  } catch (error) {
    console.error("Rule error during analysis of : " + sourceFile.fileName, error);
    return [];
  }
}

function isTypedRule(rule: tslint.IRule): rule is tslint.ITypedRule {
  return "applyWithProgram" in rule;
}
