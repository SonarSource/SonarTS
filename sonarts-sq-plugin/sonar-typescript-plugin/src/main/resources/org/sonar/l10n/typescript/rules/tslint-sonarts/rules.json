[
  {
    "key": "cognitive-complexity",
    "name": "Cognitive Complexity of functions should not be too high",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/cognitive-complexity.md"
  },
  {
    "key": "consecutive-overloads",
    "name": "Method overloads should be grouped together",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/consecutive-overloads.md"
  },
  {
    "key": "max-union-size",
    "name": "Union types should not have too many elements",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/max-union-size.md"
  },
  {
    "key": "mccabe-complexity",
    "name": "Functions should not be too complex",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/mccabe-complexity.md"
  },
  {
    "key": "no-accessor-field-mismatch",
    "name": "Getters and setters should access the expected fields",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-accessor-field-mismatch.md"
  },
  {
    "key": "no-all-duplicated-branches",
    "name": "All branches in a conditional structure should not have exactly the same implementation",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-all-duplicated-branches.md"
  },
  {
    "key": "no-array-delete",
    "name": "\"delete\" should not be used on arrays",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-array-delete.md"
  },
  {
    "key": "no-big-function",
    "name": "Functions should not have too many lines of code",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-big-function.md"
  },
  {
    "key": "no-case-with-or",
    "name": "Logical OR should not be used in switch cases",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-case-with-or.md"
  },
  {
    "key": "no-collapsible-if",
    "name": "Collapsible \"if\" statements should be merged",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-collapsible-if.md"
  },
  {
    "key": "no-collection-size-mischeck",
    "name": "Collection sizes and array length comparisons should make sense",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-collection-size-mischeck.md"
  },
  {
    "key": "no-commented-code",
    "name": "Sections of code should not be \"commented out\"",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-commented-code.md"
  },
  {
    "key": "no-dead-store",
    "name": "Dead stores should be removed",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-dead-store.md"
  },
  {
    "key": "no-duplicate-in-composite",
    "name": "Union and intersection types should not be defined with duplicated elements",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-duplicate-in-composite.md"
  },
  {
    "key": "no-duplicate-string",
    "name": "String literals should not be duplicated",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-duplicate-string.md"
  },
  {
    "key": "no-duplicated-branches",
    "name": "Two branches in a conditional structure should not have exactly the same implementation",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-duplicated-branches.md"
  },
  {
    "key": "no-element-overwrite",
    "name": "Collection elements should not be replaced unconditionally",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-element-overwrite.md"
  },
  {
    "key": "no-empty-destructuring",
    "name": "Destructuring patterns should not be empty",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-empty-destructuring.md"
  },
  {
    "key": "no-empty-nested-blocks",
    "name": "Nested blocks of code should not be left empty",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-empty-nested-blocks.md"
  },
  {
    "key": "no-extra-semicolon",
    "name": "Extra semicolons should be removed",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-extra-semicolon.md"
  },
  {
    "key": "no-gratuitous-expressions",
    "name": "Conditions should not always evaluate to \"true\" or to \"false\"",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-gratuitous-expressions.md"
  },
  {
    "key": "no-hardcoded-credentials",
    "name": "Credentials should not be hard-coded",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-hardcoded-credentials.md"
  },
  {
    "key": "no-identical-conditions",
    "name": "Related \"if/else if\" statements and \"cases\" in a \"switch\" should not have the same condition",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-identical-conditions.md"
  },
  {
    "key": "no-identical-expressions",
    "name": "Identical expressions should not be used on both sides of a binary operator",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-identical-expressions.md"
  },
  {
    "key": "no-identical-functions",
    "name": "Functions should not have identical implementations",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-identical-functions.md"
  },
  {
    "key": "no-ignored-initial-value",
    "name": "Function parameters, caught exceptions and foreach variables\u0027 initial values should not be ignored",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-ignored-initial-value.md"
  },
  {
    "key": "no-ignored-return",
    "name": "Return values should not be ignored when function calls don\u0027t have any side effects",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-ignored-return.md"
  },
  {
    "key": "no-in-misuse",
    "name": "\"in\" operator should not be used on array",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-in-misuse.md"
  },
  {
    "key": "no-inconsistent-return",
    "name": "Functions should use \"return\" consistently",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-inconsistent-return.md"
  },
  {
    "key": "no-invalid-await",
    "name": "\"await\" should only be used with promises",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-invalid-await.md"
  },
  {
    "key": "no-inverted-boolean-check",
    "name": "Boolean checks should not be inverted",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-inverted-boolean-check.md"
  },
  {
    "key": "no-misleading-array-reverse",
    "name": "Array-mutating methods should not be used misleadingly",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-misleading-array-reverse.md"
  },
  {
    "key": "no-misspelled-operator",
    "name": "Non-existent operators \u0027\u003d+\u0027, \u0027\u003d-\u0027 and \u0027\u003d!\u0027 should not be used",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-misspelled-operator.md"
  },
  {
    "key": "no-multiline-string-literals",
    "name": "Multiline string literals should not be used",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-multiline-string-literals.md"
  },
  {
    "key": "no-nested-incdec",
    "name": "Increment (++) and decrement (--) operators should not be used in a method call or mixed with other operators in an expression",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-nested-incdec.md"
  },
  {
    "key": "no-nested-template-literals",
    "name": "Template literals should not be nested",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-nested-template-literals.md"
  },
  {
    "key": "no-redundant-boolean",
    "name": "Boolean literals should not be redundant",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-redundant-boolean.md"
  },
  {
    "key": "no-redundant-jump",
    "name": "Jump statements should not be redundant",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-redundant-jump.md"
  },
  {
    "key": "no-redundant-parentheses",
    "name": "Redundant pairs of parentheses should be removed",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-redundant-parentheses.md"
  },
  {
    "key": "no-return-type-any",
    "name": "Primitive return types should be used",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-return-type-any.md"
  },
  {
    "key": "no-same-line-conditional",
    "name": "Conditionals should start on new lines",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-same-line-conditional.md"
  },
  {
    "key": "no-self-assignment",
    "name": "Variables should not be self-assigned",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-self-assignment.md"
  },
  {
    "key": "no-small-switch",
    "name": "\"switch\" statements should have at least 3 \"case\" clauses",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-small-switch.md"
  },
  {
    "key": "no-statements-same-line",
    "name": "Statements should be on separate lines",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-statements-same-line.md"
  },
  {
    "key": "no-unconditional-jump",
    "name": "Loops with at most one iteration should be refactored",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-unconditional-jump.md"
  },
  {
    "key": "no-undefined-argument",
    "name": "\"undefined\" should not be passed as the value of optional parameters",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-undefined-argument.md"
  },
  {
    "key": "no-unenclosed-multiline-block",
    "name": "Multiline blocks should be enclosed in curly braces",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-unenclosed-multiline-block.md"
  },
  {
    "key": "no-unthrown-error",
    "name": "Errors should not be created without being thrown",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-unthrown-error.md"
  },
  {
    "key": "no-unused-array",
    "name": "Array contents should be used",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-unused-array.md"
  },
  {
    "key": "no-use-of-empty-return-value",
    "name": "The output of functions that don\u0027t return anything should not be used",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-use-of-empty-return-value.md"
  },
  {
    "key": "no-useless-cast",
    "name": "Redundant casts and not-null assertions should be avoided",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-useless-cast.md"
  },
  {
    "key": "no-useless-increment",
    "name": "Values should not be uselessly incremented",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-useless-increment.md"
  },
  {
    "key": "no-useless-intersection",
    "name": "Types without members, \u0027any\u0027 and \u0027never\u0027 should not be used in type intersections",
    "type": "BUG",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-useless-intersection.md"
  },
  {
    "key": "no-variable-usage-before-declaration",
    "name": "Variables should be declared before they are used",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-variable-usage-before-declaration.md"
  },
  {
    "key": "parameters-max-number",
    "name": "Functions should not have too many parameters",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/parameters-max-number.md"
  },
  {
    "key": "prefer-default-last",
    "name": "\"default\" clauses should be last",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/prefer-default-last.md"
  },
  {
    "key": "prefer-immediate-return",
    "name": "Local variables should not be declared and then immediately returned or thrown",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/prefer-immediate-return.md"
  },
  {
    "key": "prefer-optional",
    "name": "Optional property declarations should use \u0027?\u0027 syntax",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/prefer-optional.md"
  },
  {
    "key": "prefer-promise-shorthand",
    "name": "Shorthand promises should be used",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/prefer-promise-shorthand.md"
  },
  {
    "key": "use-primitive-type",
    "name": "Wrapper objects should not be used for primitive types",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/use-primitive-type.md"
  },
  {
    "key": "use-type-alias",
    "name": "Type aliases should be used",
    "type": "CODE_SMELL",
    "url": "https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/use-type-alias.md"
  }
]