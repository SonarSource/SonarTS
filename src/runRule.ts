/*
 * SonarTS
 * Copyright (C) 2017-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as fs from "fs";
import * as path from "path";
import * as tslint from "tslint";
import { parseErrorsFromMarkup } from "tslint/lib/test/parse";
import * as ts from "typescript";

export interface PositionInFile {
  line: number;
  col: number;
}

export interface LintError {
  startPos: PositionInFile;
  endPos: PositionInFile;
  message: string;
}

export interface RuleRunResult {
  actualErrors: LintError[];
  expectedErrors: LintError[];
}

/**
 * Run rule againts a lint file
 * Use from the test: runRule(Rule, __filename) if lint file name matches the test file name
 */
export default function runRule(Rule: any, testFileName: string): RuleRunResult {
  const options: tslint.IOptions = {
    disabledIntervals: [],
    ruleArguments: [],
    ruleName: "",
    ruleSeverity: "error",
  };
  const rule = new Rule(options);
  const lintFileName = getLintFileName(testFileName);
  const { source, sourceWithFailures } = getSources(lintFileName);
  const sourceFile = tslint.getSourceFile(testFileName, source);
  const failures = rule.apply(sourceFile);
  const actualErrors = mapToLintErrors(failures);
  const expectedErrors = parseErrorsFromMarkup(sourceWithFailures);
  return { actualErrors, expectedErrors };
}

function getLintFileName(testFileName: string) {
  const baseName = path.basename(testFileName, ".test.ts");
  return path.join(__dirname, `./rules/${baseName}.lint`);
}

function getSources(fileName: string) {
  const sourceWithFailures = fs.readFileSync(fileName, "utf-8");
  const source = sourceWithFailures
    .split("\n")
    .filter((line) => line.indexOf("~") === -1)
    .join("\n");
  return { source, sourceWithFailures };
}

function mapToLintErrors(failures: tslint.RuleFailure[]) {
  return failures.map((failure) => {
    const startPosition = failure.getStartPosition().getLineAndCharacter();
    const endPosition = failure.getEndPosition().getLineAndCharacter();
    return {
      endPos: { col: endPosition.character, line: endPosition.line },
      message: failure.getFailure(),
      startPos: { col: startPosition.character, line: startPosition.line },
    };
  });
}
