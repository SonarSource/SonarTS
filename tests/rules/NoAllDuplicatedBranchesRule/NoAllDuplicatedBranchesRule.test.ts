import { Rule } from "../../../src/rules/NoAllDuplicatedBranchesRule";
import { runRule, runRuleOnRuling } from "../../runRule";

it("raises error", () => {
  const { actualErrors, expectedErrors } = runRule(Rule, __filename);
  expect(actualErrors).toEqual(expectedErrors);
});

it("ruling", () => {
  expect(runRuleOnRuling(Rule)).toMatchSnapshot();
});
