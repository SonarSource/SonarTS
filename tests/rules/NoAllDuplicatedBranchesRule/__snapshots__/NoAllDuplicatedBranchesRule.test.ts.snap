// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ruling 1`] = `
Array [
  "src/angular/packages/compiler-cli/src/codegen.ts: 47",
  "src/vscode/src/vs/editor/browser/view/viewPart.ts: 47",
  "src/vscode/src/vs/editor/contrib/comment/common/lineCommentCommand.ts: 122",
  "src/vscode/src/vs/editor/contrib/dnd/common/dragAndDropCommand.ts: 85",
]
`;
